openscm/timeframes.py

```diff
@@ -0,0 +1,513 @@
+    return np.concatenate(
+        (
+            np.array(
+                [
+                    np.concatenate(([first_edge_point], edge_points)),
+                    np.concatenate(([values[0]], middle_points, [values[-1]])),
+                ]
+            ).T.reshape(2 * len(values)),
+            [last_edge_point],
+        )
+    )
+
+
+def _calc_linearization(
+    values: np.ndarray, timeframe: Timeframe
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Calculate the "linearization" time points and values (see
+    :func:`openscm.timeframes._calc_linearization_values`) of the array ``values``
+    according to timeframe ``timeframe``.
+
+    Parameters
+    ----------
+    values
+        Timeseries values of period averages
+    timeframe
+        Timeframe
+
+    Returns
+    -------
+    linearization_points
+        Linearization time points ("x-values")
+    linearization_values
+        Linearization values ("y-values")
+    """
+    stop_time = timeframe.start_time + len(values) * timeframe.period_length
+    linearization_points = np.linspace(
+        timeframe.start_time, stop_time, 2 * len(values) + 1
+    )
+    linearization_values = _calc_linearization_values(values)
+    return linearization_points, linearization_values
+
+
+class _Interpolation(NamedTuple):
+    """
+    Collection of interpolation points and associated information.
+    """
+
+    data_len: int
+    """
+    Length of source data array for which these interpolation points have been
+    calculated
+    """
+
+    first_interval_length: int
+    """
+    Length of the first interval/period (can be smaller than ``period_length``
+    when not enough data overlap is available)
+    """
+
+    last_interval_length: int
+    """
+    Length of the last interval/period (can be smaller than ``period_length``
+    when not enough data overlap is available)
+    """
+
+    points: np.ndarray
+    """Array of interpolation time points"""
+
+    period_length: int
+    """
+    Length of the middle intervals/periods
+    """
+
+    target_indices: np.ndarray
+    """
+    Array of indices in ``interpolation_points`` that correspond to edges of target
+    periods
+    """
+
+
+def _calc_interpolation_points(
+    source_len: int, source: Timeframe, target: Timeframe
+) -> _Interpolation:
+    """
+    Calculate the "interpolation" time points that correspond to edges of the target
+    periods and the sampling points for the linearization of source values (i.e. edges
+    and middle points of the source periods).
+
+    Parameters
+    ----------
+    source_len
+        Length of source data values
+    source
+        Source timeframe
+    target
+        Target timeframe
+
+    Returns
+    -------
+    _Interpolation
+        Object wrapping the actual interpolation points and additional information
+    """
+    linearization_points_len = 2 * source_len + 1
+    # Use double the time points to stick to integers when using half the source period
```

Comment:
What does this mean?

Reply:
I tried to clarify that in the updated comment